### PR TITLE
Fix image-related property changes not being detected

### LINKED ISSUE #7288: [Android] Update/Binding on BitmapImage does not update parent ImageBrush
### Current behavior

When using a BitmapImage as ImageSource for a ImageBrush, updates on the BitmapImage are not triggering updates on the brush.

Sample:
```
        <Grid x:Name="tile">
            <Border Width="48" Height="48" CornerRadius="8">
                <Border.Background>
                    <ImageBrush Stretch="UniformToFill" x:Name="brush">
                        <ImageBrush.ImageSource>
                            <BitmapImage UriSource="{x:Bind TileContent, Mode=OneWay}" DecodePixelType="Logical" DecodePixelWidth="48" DecodePixelHeight="48"/>
                        </ImageBrush.ImageSource>
                    </ImageBrush>
                </Border.Background>
            </Border>
        </Grid>
```

In the sample above, the UriSource is bound against a property of type Uri. Whenever the uri changes, the new image shall be loaded and the content of the ImageBrush should change.

While this works as expected under Windows, on Android only the initial BitmapImage will be created and any further change (via binding or set programmatically) will be ignored.
Obviously the ImageBrush does not realize the actual instance of its BitmapImage changed.

### Expected behavior

Should behave like under Windows. Any change to the BitmapImage (e.g. by modifying the UriSource property) should also be reflected by the ImageBrush using the BitmapImage as its ImageSource.

### How to reproduce it (as minimally and precisely as possible)

Use the snippet above and bind against a DependencyProperty. Switching the uri will change the image under Windows, while it will be created under Android once and not change afterwards.

### Workaround

Instead of modifying the BitmapImage, under Android the ImageSource has to be set explicitly.
In my case I added code-behind:

```
        public MyControl()
        {
            this.InitializeComponent();
            this.RegisterPropertyChangedCallback(TileContentProperty, TileContentChanged);
        }

        private void TileContentChanged(DependencyObject sender, DependencyProperty dp)
        {
            BitmapImage bi = new BitmapImage();
            bi.DecodePixelType = DecodePixelType.Logical;
            bi.DecodePixelHeight = 48;
            bi.DecodePixelWidth = 48;
            bi.UriSource = LogoUri;

            brush.ImageSource = bi;
        }
```

Explicitly setting the ImageSource as a whole will make the ImageBrush realize that something on the underlying BitmapImage object changed.

### Works on UWP/WinUI

Yes

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia

### NuGet package version(s)

Uno.UI version 3.10.11

### Affected platforms

Android

### IDE

Visual Studio 2019

### IDE version

16.11.4

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### LINKED ISSUE #9533: `ImageOpened` and `ImageFailed` should be supported everywhere
## What would you like to be added:

Currently `ImageOpened` and `ImageFailed` events are not supported on all platforms. This is a problem as they are often used to allow asynchronous (animated) loading of network images.

## Why is this needed:

Notifying subscribers that image is loaded/loading failed.

## For which Platform:

All

### LINKED ISSUE #11673: [Droid] BitmapImage doesn't call any related event
### Current behavior

BitmapImage on android doesn't call ImageOpened, ImageFailed, DownloadProgress

### Expected behavior

Called events on image load or fail

### How to reproduce it (as minimally and precisely as possible)

Repro

1. Load image by url from text field in repro
2. No output in debug TextBlock

[BitmapImageUnoIssues.zip](https://github.com/unoplatform/uno/files/10961642/BitmapImageUnoIssues.zip)

### Workaround

_No response_

### Works on UWP/WinUI

Yes

### Environment

_No response_

### NuGet package version(s)

_No response_

### Affected platforms

Android

### IDE

Visual Studio 2022

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

Better to check iOS, MAC, Wasm too

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #11673, closes #7288, closes #9533

## PR Type

What kind of change does this PR introduce?

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Comment by Youssef1313** [comment]
I avoided any heavy refactoring in this PR, and it doesn't yet handle everything perfectly.
I'll work on handling more cases later in another PR.

File: src/Uno.UI/UI/Xaml/Media/ImageBrush.cs [hunk 1]
```diff
@@ -65,14 +69,32 @@ public Stretch Stretch
...
+				_sourceDisposable.Disposable = null;
+			}
+			else if (newValue is BitmapImage bitmapImage)
+			{
+				_sourceDisposable.Disposable = bitmapImage.RegisterDisposablePropertyChangedCallback(
+					BitmapImage.UriSourceProperty, (_, _) =>
+					{
+						OnSourceChangedPartial(newValue, null);
+					});
+			}
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/Media/ImageBrush.cs:93]: if we transition from BitmapImage to something not listed here(eg: SvgImageSource), the old one isnt disposed
you should always dispose, and then set a new one if possible:
```suggestion
			_sourceDisposable.Disposable = null;
			if (newValue is BitmapImage bitmapImage)
			{
				_sourceDisposable.Disposable = bitmapImage.RegisterDisposablePropertyChangedCallback(
					BitmapImage.UriSourceProperty, (_, _) =>
					{
						OnSourceChangedPartial(newValue, null);
					});
			}
```

File: src/Uno.UI/UI/Xaml/Media/Brush.Android.cs [hunk 2]
```diff
@@ -93,6 +94,22 @@ internal static IDisposable AssignAndObserveBrush(Brush b, ColorSetterHandler co
 				var disposables = new CompositeDisposable(5);
 				imageBrushCallback();
 
+				var imageBrushSource = imageBrush.ImageSource;
+				if (imageBrushSource is SvgImageSource svgImageSource)
+				{
+					svgImageSource.RegisterDisposablePropertyChangedCallback(
+						SvgImageSource.UriSourceProperty,
+						(_, _) => imageBrushCallback()
+					).DisposeWith(disposables);
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/Media/Brush.Android.cs:103]:
```suggestion
					svgImageSource
						.RegisterDisposablePropertyChangedCallback(SvgImageSource.UriSourceProperty, ImageBrushCallback)
						.DisposeWith(disposables);
```

Comment (Xiaoy312) [reply src/Uno.UI/UI/Xaml/Media/Brush.Android.cs:103]: and define this local function:
```cs
void ImageBrushCallback(DependencyObject sender, DependencyPropertyChangedEventArgs e) => imageBrushCallback();
```
and, you can replace rest of lambda funcs too

File: src/Uno.UI/UI/Xaml/Controls/Image/Image.cs [hunk 3]
```diff
@@ -56,13 +56,20 @@ public static new class TraceProvider
...
 			}
 
+#if !__SKIA__ && !__WASM__ // TODO: Have consistent handling on Wasm and Skia.
+			if (imageSource is BitmapImage bitmapImage && exception is not null)
+			{
+				bitmapImage.RaiseImageFailed(exception);
+			}
+#endif
+
 			ImageFailed?.Invoke(this, new ExceptionRoutedEventArgs(this, "Image failed to download"));
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/Controls/Image/Image.cs:73]: more context is always better:
```suggestion
			ImageFailed?.Invoke(this, new ExceptionRoutedEventArgs(this, "Image failed to download: " + exception));
```

File: src/Uno.UI/UI/Xaml/Controls/Image/Image.cs [hunk 4]
```diff
@@ -56,13 +56,20 @@ public static new class TraceProvider
 			public const int Image_SetImageStop = 6;
 		}
 
-		protected virtual void OnImageFailed(ImageSource imageSource)
+		private protected void OnImageFailed(ImageSource imageSource, Exception exception)
 		{
 			if (this.Log().IsEnabled(Uno.Foundation.Logging.LogLevel.Debug))
 			{
 				this.Log().Debug(this.ToString() + " Image failed to open");
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/Controls/Image/Image.cs:63]: more context is always better:
```suggestion
				this.Log().Debug($"Image '{this}' failed to open: {exception}");
```

File: src/Uno.UI/UI/Xaml/Controls/Image/Image.cs [hunk 5]
```diff
@@ -56,14 +56,35 @@ public static new class TraceProvider
 			public const int Image_SetImageStop = 6;
 		}
 
-		protected virtual void OnImageFailed(ImageSource imageSource)
+		private protected void OnImageFailed(ImageSource imageSource, Exception exception)
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/Controls/Image/Image.cs:59]: we could just avoid the null check... there is little point to do a branching here

File: src/Uno.UI/UI/Xaml/Media/ImageBrush.cs [hunk 6]
```diff
@@ -65,14 +69,32 @@ public Stretch Stretch
...
 
+		private void OnSourceChanged(ImageSource newValue, ImageSource oldValue)
+		{
+			if (newValue is BitmapImage bitmapImage)
+			{
+				_sourceDisposable.Disposable = bitmapImage.RegisterDisposablePropertyChangedCallback(
+					BitmapImage.UriSourceProperty, (_, _) =>
+					{
+						OnSourceChangedPartial(newValue, null);
+					});
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/Media/ImageBrush.cs:88]:
```suggestion
				_sourceDisposable.Disposable = bitmapImage.RegisterDisposablePropertyChangedCallback(
					BitmapImage.UriSourceProperty,
					(_, _) => OnSourceChangedPartial(newValue, null)
				);
```

**Review by Xiaoy312 — APPROVED** [review]

File: src/Uno.UI/UI/Xaml/Controls/Image/Image.cs (hunk 5 above)
Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/Image/Image.cs:59]: @Xiaoy312 I'm quite afraid `exception` could be null in this code path

**Comment by Youssef1313** [comment]
@jeromelaban @Xiaoy312 Is this ready to merge?

File: src/Uno.UI/UI/Xaml/Controls/Image/Image.cs (hunk 5 above)
Comment (Xiaoy312) [reply src/Uno.UI/UI/Xaml/Controls/Image/Image.cs:59]: $"It is perfectly fine to print a null exception: {(Exception)null}" // for the logging parts

**Pull request merged**